src/react/utils.js

```diff
@@ -543,26 +543,38 @@ export function getComponentTypeFromRootValue(realm: Realm, value: Value): ECMAS
   }
 }
 
-// props should never have "ref" or "key" properties, as they're part of ReactElement
-// object instead. to ensure that we can give this hint, we create them and then
-// delete them, so their descriptor is left undefined. we use this knowledge later
-// to ensure that when dealing with creating ReactElements with partial config,
-// we don't have to bail out becuase "config" may or may not have "key" or/and "ref"
-export function deleteRefAndKeyFromProps(realm: Realm, props: ObjectValue | AbstractObjectValue): void {
-  setProperty(props, "ref", realm.intrinsics.undefined);
-  deleteProperty(props, "ref");
-  setProperty(props, "key", realm.intrinsics.undefined);
-  deleteProperty(props, "key");
-}
-
-export function objectHasNoPartialKeyAndRef(
-  realm: Realm,
-  object: ObjectValue | AbstractValue | AbstractObjectValue
-): boolean {
-  if (object instanceof AbstractValue) {
+export function flagPropsWithNoPartialKeyOrRef(realm: Realm, props: ObjectValue | AbstractObjectValue): void {
+  realm.react.propsWithNoPartialKeyOrRef.add(props);
```

Comment:
Can/should we assert that the passed object is final? So we don't pass something non-final by mistake.

Reply:
Unfortunately not, abstract object values cannot be final – only their underlying template can be and not all abstract object values have templates (that's why we do the top check).

Reply:
Can we at least check finality for cases where it's not abstract (or when we have all templates)?